src/Polly.Core/Utils/Pipeline/CompositeComponent.cs

```diff
@@ -128,25 +128,44 @@ private sealed class DelegatingComponent : PipelineComponent
 
         public PipelineComponent? Next { get; set; }
 
+        public override ValueTask DisposeAsync() => default;
+
         internal override ValueTask<Outcome<TResult>> ExecuteCore<TResult, TState>(
             Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
             ResilienceContext context,
             TState state)
         {
+            // Custom state object is used to cast the callback and state to prevent infinite
+            // generic type recursion warning IL3054 when referenced in a native AoT application.
+            // See https://github.com/App-vNext/Polly/issues/1732 for further context.
             return _component.ExecuteCore(
-                static (context, state) =>
+                static (context, wrapper) =>
                 {
+                    var callback = (Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>>)wrapper.Callback;
+                    var state = (TState)wrapper.State;
                     if (context.CancellationToken.IsCancellationRequested)
                     {
                         return Outcome.FromExceptionAsValueTask<TResult>(new OperationCanceledException(context.CancellationToken).TrySetStackTrace());
                     }
 
-                    return state.Next!.ExecuteCore(state.callback, context, state.state);
+                    return wrapper.Next.ExecuteCore(callback, context, state);
                 },
                 context,
-                (Next, callback, state));
+                new StateWrapper(Next!, callback, state!));
```

Comment:
Note: this allocation happens for every strategy in that pipeline.

So 5 strategies --> around 100 bytes allocated per execution.